### PR TITLE
[DRAFT] [models] add ViTSTR in TF and PT

### PR DESCRIPTION
**Pull request by felixdittrich92**
This PR:

- adds VisionTransformer as module in models/modules/vision_transformer (TF/PT) 
- add Vitstr head (TF / PT)

TODOS: 

- [ ] toy run / check that all works fine
- [ ] check code quality

Any feedback is welcome :)

@frgfm wdyt ? I think it is a more flexible way instead of adding ViT as a fixed classification model (where we run into trouble with different sizes for patch_embeds) - can easily be extended needs only a head :sweat_smile: 

related to: #513 #1003

### CODE REVIEW

**Comment by felixdittrich92** [comment]
NOTE: 
- ViT works fine (also tested with timm's ViT implementation)
- loss does not decrease (needs debugging)

**Comment by felixdittrich92** [comment]
@frgfm before i continue to debug this let me know your ideas you have in mind :hugs: 
Go this way to keep ViT as module or maybe start to implement as classification model and then think how we can make the patch_embedding flexible ? wdyt ?

File: doctr/models/modules/transformer/pytorch.py [hunk 1]
```diff
@@ -57,10 +57,10 @@ def scaled_dot_product_attention(
 class PositionwiseFeedForward(nn.Sequential):
     """Position-wise Feed-Forward Network"""
 
-    def __init__(self, d_model: int, ffd: int, dropout: float = 0.1) -> None:
+    def __init__(self, d_model: int, ffd: int, dropout: float = 0.1, use_gelu: bool = False) -> None:
```
Comment (frgfm) [inline doctr/models/modules/transformer/pytorch.py:60]: boolean value for activation selection is rather limited: are we positive that only relu & gelu can be used for such architecture types?

File: doctr/models/modules/transformer/tensorflow.py [hunk 2]
```diff
@@ -74,14 +74,19 @@ def scaled_dot_product_attention(
 class PositionwiseFeedForward(layers.Layer, NestedObject):
     """Position-wise Feed-Forward Network"""
 
-    def __init__(self, d_model: int, ffd: int, dropout=0.1) -> None:
+    def __init__(self, d_model: int, ffd: int, dropout=0.1, use_gelu: bool = False) -> None:
         super(PositionwiseFeedForward, self).__init__()
+        self.use_gelu = use_gelu
```
Comment (frgfm) [inline doctr/models/modules/transformer/tensorflow.py:79]: instantiate a `self.activation_fn` in the constructor to avoid conditional execution in the call :+1:

File: doctr/models/modules/vision_transformer/pytorch.py [hunk 3]
```diff
@@ -0,0 +1,100 @@
...
+
+from ..transformer.pytorch import MultiHeadAttention, PositionwiseFeedForward
+
+__all__ = ["VisionTransformer"]
+
+
+class PatchEmbedding(nn.Module):
+    """Compute 2D patch embedding"""
+
+    # Inpired by: https://github.com/rwightman/pytorch-image-models/blob/master/timm/models/layers/patch_embed.py
```
Comment (frgfm) [inline doctr/models/modules/vision_transformer/pytorch.py:19]: Just FYI: can you confirm that you made a lot of modifications?
inspired by is rather light
borrowed from is more significant

File: doctr/models/modules/vision_transformer/pytorch.py [hunk 4]
```diff
@@ -0,0 +1,100 @@
...
+        assert H % self.patch_size[0] == 0, "Image height must be divisible by patch height"
+        assert W % self.patch_size[1] == 0, "Image width must be divisible by patch width"
+
+        return self.proj(x)  # BCHW
+
+
+class VisionTransformer(nn.Module):
+    """VisionTransformer architecture as described in
+    `"An Image is Worth 16x16 Words: Transformers for Image Recognition at Scale",
+    <https://arxiv.org/pdf/2010.11929.pdf>`_."""
```
Comment (frgfm) [inline doctr/models/modules/vision_transformer/pytorch.py:47]: let's specify the constructor args

File: doctr/models/modules/vision_transformer/tensorflow.py [hunk 5]
```diff
@@ -0,0 +1,110 @@
...
+        assert H % self.patch_size[0] == 0, "Image height must be divisible by patch height"
+        assert W % self.patch_size[1] == 0, "Image width must be divisible by patch width"
+
+        return self.proj(x, **kwargs)  # BHWC
+
+
+class VisionTransformer(layers.Layer, NestedObject):
+    """VisionTransformer architecture as described in
+    `"An Image is Worth 16x16 Words: Transformers for Image Recognition at Scale",
+    <https://arxiv.org/pdf/2010.11929.pdf>`_."""
```
Comment (frgfm) [inline doctr/models/modules/vision_transformer/tensorflow.py:54]: same here

**Review by frgfm — COMMENTED** [review]
Thanks Felix! I added some comments :)

**Comment by felixdittrich92** [comment]
@frgfm haha i do currently focus to implement it as classification model :sweat_smile: we have to do a decision !

**Comment by felixdittrich92** [comment]
outdated by: #1050
